### PR TITLE
[DT 1459] Naming service controller and get domain

### PR DESCRIPTION
**Pull request by tunguyennnnn**
## Background

Refactor to have multiple blockchain + naming services 

## Changes
* Add new naming service controller class
* Apply the new naming service controller to get domains/:domain controller

## Code Review

This Pull Request was thoroughly reviewed and meets all Code Review Standards pertaining to:

- [ ] **Implementation** - satisfied acceptance criteria
- [ ] **Communication** - notified engineers/stakeholders about impactful changes
- [ ] **Error handling** - handled, logged & alerted on errors
- [ ] **Documentation** - wrote readable code & commits, documented unintuitive code
- [ ] **Testing** - unit & E2E test coverage, tested in staging, DB migrations backwards compatible
- [ ] **Security** - sanitized inputs, vetted dependencies, validated/secured API requests, no committed secrets
- [ ] **Performance** - optimized expensive algorithms & database queries

Please select all **applied** standards relevant to this PR.
## Confirmation

- [ ] **Assignee Confirmation** - I (the author) have filled out the template above
- [ ] **Reviewer Confirmation** - I (a/the reviewer) confirm 1) the author has filled out the template and checked the box that says **Assignee Confirmation** 2) I reviewed the code and selected all applicable items in the code review section.

### CODE REVIEW

File: src/services/nameing-services/EnsNamingService.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import { ensNameHash } from '../../utils/namehash';
+import { EnsDomainSchema } from '../../utils/validators/domain-validators/schema';
+import { domainsBatchLoader } from '../DomainService';
+import {
+  getDomainResolution,
+  resolutionRecordsToKeyValue,
+} from '../Resolution';
+import { NamingService } from './BaseNamingService';
+
+resolutionRecordsToKeyValue;
```
Comment (sammyluo) [inline src/services/nameing-services/EnsNamingService.ts:15]: ?

**Review by sammyluo — CHANGES_REQUESTED** [review]
Nice refactoring! Great start! 🙌 Just left some minor questions and change requests.

File: src/utils/evmUtils.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+export const ETHAddressRegex = /^0x[a-fA-F0-9]{40}$/;
+export const NullAddress = '0x0000000000000000000000000000000000000000';
+export const VALID_TOKEN_REGEX = /^[a-fA-F0-9]+$/;
+
+export const getTokenIdFromHash = (hash: string): string => {
+  return BigNumber.from(hash).toString();
+};
+
+export const isNullAddress = (address: string | null): boolean => {
+  return address === null || address === NullAddress;
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/utils/evmUtils.ts:12]: I think that would be useful for another function but i've seen use case for it, but need a different function signature:
```
export const isEthNullAddress = (address: string): boolean =>
```

File: src/middleware/inputValidators/validateDomainNameQuery.ts [hunk 3]
```diff
@@ -77,3 +142,31 @@ export function ValidateAndTransformOnDomainName(
...
+    let value = req.params[param].trim();
+
+    if (!value) {
+      throw new InvalidInputError('Empty input');
+    }
+
+    const namingService = getNamingServiceForDomain(value);
+
+    if (!namingService) {
+      throw new InvalidInputError('Unsupported TLD');
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/middleware/inputValidators/validateDomainNameQuery.ts:159]: i think this should be job of middleware and controller to throw custom Error
the `InvalidInputError` is a custom http error with code, getNamingServiceForDomain concern should only care if there's a service to handle it

File: src/services/nameing-services/UdNamingService.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+  getNameHash(domain: string): string {
+    return eip137Namehash(domain);
+  }
+
+  isOwnerAddressValid(address: string): boolean {
+    return ETHAddressRegex.test(address);
+  }
+
+  protected buildDomainResponse(
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/services/nameing-services/UdNamingService.ts]: yea agree! I was thinking of another custom return type in `GetDomainResponse` 
The reason why we may not return `Domain` is if we have API proxy for another naming service, it doesn't fit into `Domain` model
Let's figure out when we reach to that point though, for now all namingservice return the same thing

File: src/middleware/InputValidatorsMiddleware.test.ts [hunk 5]
```diff
@@ -201,4 +205,242 @@ describe('InputValidatorMiddleware', () => {
...
+    it('should pass for valid name hash', () => {
+      mockRequest = {
+        params: {
+          domainOrToken: '0x1234567890',
+        },
+      } as Request & { params: any; locals: any };
+
+      middleware(mockRequest, mockResponse, next);
+      expect(next.calledOnce);
+      expect(mockResponse.locals.namingService instanceof ZilNamingService).to
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/middleware/InputValidatorsMiddleware.test.ts:427]: yea this should not have any naming service 😄

File: src/services/nameing-services/index.ts [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  supportedResolveKeys: Object.keys(SupportedKeysJson.keys),
+});
+
+export const getNamingServiceForDomain = (
+  domain: string,
+): NamingService | null => {
+  const { extension } = splitDomain(domain);
+
+  if (!extension || !SupportedTlds.includes(extension)) {
+    return null;
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/services/nameing-services/index.ts:35]: yea lets the controller + middleware throw and decide what error is it instead

File: src/services/nameing-services/EnsNamingService.ts (hunk 1 above)
Comment (tunguyennnnn) [reply src/services/nameing-services/EnsNamingService.ts:15]: 🔪

File: src/middleware/InputValidatorsMiddleware.test.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/middleware/InputValidatorsMiddleware.test.ts:427]: FYI, for tokenId we cannot determine naming service by domain name, so i defer this to MetadataController to handle which naming service to handle

File: src/services/nameing-services/ZilNamingService.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    const parsedDomain = UnsDomainSchema.parse(domain);
+    return parsedDomain;
+  }
+
+  getNameHash(domain: string): string {
+    return eip137Namehash(domain);
+  }
+
+  private async getDomainsByName(domainName: string): Promise<Domain[]> {
+    const udNameHash = this.getNameHash(domainName);
```
Comment (sammyluo) [inline src/services/nameing-services/ZilNamingService.ts:25]: Nit: 
```suggestion
    const unsNameHash = this.getNameHash(domainName);
```

**Review by sammyluo — APPROVED** [review]
LGTM! 🚀 ⭐

**Review by DancingAxolotl — APPROVED** [review]
Great work! One minor comment, otherwise LGTM 🙌 🚀

File: src/services/nameing-services/BaseNamingService.ts [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+   */
+  abstract isOwnerAddressValid(address: string): boolean;
+
+  async getDomainByName(_: string): Promise<DomainResponse | null> {
+    return null;
+  }
+
+  async getDomainFromTokenId(tokenId: string): Promise<Domain | null> {
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/services/nameing-services/BaseNamingService.ts:119]: ah yea they are abstract methods too for controller 😁

[the remark this answers is not in the document]
Comment (tunguyennnnn) [reply src/services/nameing-services/BaseNamingService.ts:119]: ah yea  for `getDomainByName`
the `getDomainFromTokenId` is not implemented by Uns naming service yet so I'd leave it to return null for now

**Pull request merged**